yb-voyager/cmd/importData.go

```diff
@@ -968,69 +968,98 @@ func splitFilesForTable(state *ImportDataState, filePath string, t sqlname.NameT
-		line, readLineErr = dataFile.NextLine()
+		line, currentBytesRead, readLineErr = dataFile.NextLine()
 		if readLineErr == nil || (readLineErr == io.EOF && line != "") {
 			// handling possible case: last dataline(i.e. EOF) but no newline char at the end
 			numLinesTaken += 1
 		}
+		log.Debugf("Batch %d: totalBytesRead %d, currentBytes %d \n", batchNum, dataFile.GetBytesRead(), currentBytesRead)
+		if currentBytesRead >= tdb.MaxBatchSizeInBytes() {
+			//If a row is itself larger than MaxBatchSizeInBytes erroring out
+			utils.ErrExit("record num=%d for table %q in file %s is larger than %d bytes batch size", numLinesTaken, t.ForOutput(), filePath, tdb.MaxBatchSizeInBytes())
+		}
 		if line != "" {
 			// can't use importBatchArgsProto.Columns as to use case insenstiive column names
 			columnNames, _ := TableToColumnNames.Get(t)
 			line, err = valueConverter.ConvertRow(t, columnNames, line)
 			if err != nil {
-				utils.ErrExit("transforming line number=%d for table %q in file %s: %s", batchWriter.NumRecordsWritten+1, t, filePath, err)
+				utils.ErrExit("transforming line number=%d for table %q in file %s: %s", numLinesTaken, t.ForOutput(), filePath, err)
 			}
-		}
-		err = batchWriter.WriteRecord(line)
-		if err != nil {
-			utils.ErrExit("Write to batch %d: %s", batchNum, err)
-		}
-		if batchWriter.NumRecordsWritten == batchSize ||
-			dataFile.GetBytesRead() >= tdb.MaxBatchSizeInBytes() ||
-			readLineErr != nil {
 
-			isLastBatch := false
-			if readLineErr == io.EOF {
-				isLastBatch = true
-			} else if readLineErr != nil {
-				utils.ErrExit("read line from data file %q: %s", filePath, readLineErr)
+			// Check if adding this record exceeds the max batch size
+			if batchWriter.NumRecordsWritten == batchSize || 
+				dataFile.GetBytesRead() >= tdb.MaxBatchSizeInBytes() { // GetBytesRead - returns the total bytes read until now including the currentBytesRead 
+
+				// Finalize the current batch without adding the record
+				finalizeBatch(false, numLinesTaken-1, dataFile.GetBytesRead() - currentBytesRead)
+
+				//carry forward the bytes to next batch
+				dataFile.ResetBytesRead(currentBytesRead)
+
+				// Start a new batch by calling the initBatchWriter function
+				initBatchWriter()
 			}
 
-			offsetEnd := numLinesTaken
-			batch, err := batchWriter.Done(isLastBatch, offsetEnd, dataFile.GetBytesRead())
+			// Write the record to the new or current batch
+			err = batchWriter.WriteRecord(line)
 			if err != nil {
-				utils.ErrExit("finalizing batch %d: %s", batchNum, err)
+				utils.ErrExit("Write to batch %d: %s", batchNum, err)
 			}
-			batchWriter = nil
-			dataFile.ResetBytesRead()
-			submitBatch(batch, updateProgressFn, importBatchArgsProto)
+		}
 
-			if !isLastBatch {
-				batchNum += 1
-			}
+		// Finalize the batch if it's the last line or the end of the file and reset the bytes read to 0
+		if readLineErr == io.EOF {
+			finalizeBatch(true, numLinesTaken, dataFile.GetBytesRead())
+			dataFile.ResetBytesRead(0)
```

Comment:
this if-else condition is inside another condition when `line!=""` only so I believe the `readLineErr == io.EOF` can happen irrespective of line is empty or not so that's why is outside it.